Add `--debug` option to `briefcase run` (#2173)
beeware/briefcase: docs/how-to/debugging/console.rst

=================

> Debug via Console

Reviewer: We may have other console debuggers in the future, so should we change this to "PDB"?

Author: Changed to PDB

=================

Debugging an app on the console is normally done via `PDB <https://docs.python.org/3/library/pdb.html>`_.

> It is possible to debug a briefcase app at different stages in your development
> process. You can debug a development app via ``briefcase dev``, but also an bundled
> app that is build via ``briefcase build`` and run via ``briefcase run``.

Reviewer: ```suggestion
It is possible to debug a Briefcase app at different stages in your development
process. You can debug a development app via ``briefcase dev``, but also a bundled
app that is built via ``briefcase build`` and run via ``briefcase run``.
```

Author: Fixed.

Development
-----------

> Debugging an development app is quiet easy. Just add ``breakpoint()`` inside
> your code and start the app via ``briefcase dev``. When the breakpoint got hit
> the pdb console opens on your console and you can debug your app.

Reviewer: ```suggestion
Debugging an development app is quite easy. Just add ``breakpoint()`` inside
your code and start the app via ``briefcase dev``. When the breakpoint is hit,
the pdb console opens in your terminal and you can debug your app.
```

Bundled App
-----------
It is also possible to debug a bundled app. This is currently still an **experimental feature** that is only
supported on Windows and macOS. The full power of this feature will become available when iOS and

> Android are supported, because that is the only way to debug an iOS or Android app.

Reviewer: Let's not mention Android and iOS yet, because we don't know how long it'll be before they're supported.

Also, some of the lines in this file are much longer than 88 characters.

Author: Removed Android and iOS and fixed all lines longer than 88 characters.

To debug a bundled app a piece of the debugger has to be embedded into your app. This is done via:

.. code-block:: console

    $ briefcase build --debug pdb

This will build your app in debug mode and add `remote-pdb <https://pypi.org/project/remote-pdb/>`_
together with a package that automatically starts ``remote-pdb`` at the
startup of your bundled app.

Then it is time to run your app. You can do this via:

.. code-block:: console

    $ briefcase run --debug pdb

Running the app in debug mode will automatically start the ``remote-pdb`` debugger
and wait for incoming connections. By default it will listen on ``localhost``
and port ``5678``.

> Then it is time to create a new console window on your host system and connect
> to your bundled app.

Reviewer: ```suggestion
In a separate console window on your host system, connect
to your bundled app:
```

.. tabs::

  .. group-tab:: Windows
    To connect to your application, you need access to ``telnet``. That is not activated by default, but can be
    activated by running the following command with admin rights

    .. code-block:: console

      $ dism /online /Enable-Feature /FeatureName:TelnetClient

    Then you can start the connection via

    .. code-block:: console

      $ telnet localhost 5678

  .. group-tab:: Linux

    .. code-block:: console

      $ nc localhost 5678

  .. group-tab:: macOS

    .. code-block:: console

      $ nc localhost 5678

> The app will start after the connection is established.

Reviewer: Let's add a link at the end to the `briefcase run` page where the user can find more information.

Author: Added a link.

together with a package that automatically starts ``remote-pdb`` at the
startup of your bundled app.

Then it is time to run your app. You can do this via:

.. code-block:: console

    $ briefcase run --debug pdb

Running the app in debug mode will automatically start the ``remote-pdb`` debugger
and wait for incoming connections. By default it will listen on ``localhost``
and port ``5678``.

Then it is time to create a new console window on your host system and connect
to your bundled app.

.. tabs::

  .. group-tab:: Windows
    To connect to your application, you need access to ``telnet``. That is not activated by default, but can be
    activated by running the following command with admin rights

    .. code-block:: console

      $ dism /online /Enable-Feature /FeatureName:TelnetClient

    Then you can start the connection via

    .. code-block:: console

      $ telnet localhost 5678

  .. group-tab:: Linux
    If not already done install ``rlwrap`` and ``socat``

    .. code-block:: console

      $ sudo apt install rlwrap socat

    Then you can start the connection via

    .. code-block:: console

      $ rlwrap socat - tcp:localhost:5678

  .. group-tab:: macOS

>     If not already done install ``rlwrap`` and ``socat`` using `Homebrew <https://brew.sh>`__

Author: Fixed. See [here](https://github.com/beeware/briefcase/pull/2173/files/092f356227c02a2209af667b1d5f92f467fb1cb0#r2364366492).
